File: next/dweb-browser/src/Platforms/iOS/WebModule/Js/JsProcessWebApi.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using DwebBrowser.DWebView;
using DwebBrowser.IpcWeb;
using DwebBrowser.MicroService;
using DwebBrowser.MicroService.Message;
using Foundation;
using static DwebBrowser.WebModule.Js.JsProcessWebApi;

namespace DwebBrowser.WebModule.Js;

public class JsProcessWebApi
{
    public DWebView.DWebView DWebView { get; init; }

    public JsProcessWebApi(DWebView.DWebView dWebView)
    {
        DWebView = dWebView;
    }


    public class ProcessInfo
    {
        [JsonPropertyName("process_id")]
        public int ProcessId { get; set; }
        public ProcessInfo(int process_id)
        {
            ProcessId = process_id;
        }
    }

    public record ProcessHandler(ProcessInfo Info, MessagePortIpc Ipc);

    public async Task<ProcessHandler> CreateProcess(
        string env_script_url,
        string metadata_json,
        string env_json,
        Ipc.MicroModuleInfo remoteModule,
        string host)
    {
        var channel = await this.DWebView.CreateWebMessageChannel();
        var port1 = channel.Port1;
        var port2 = channel.Port2;

        var nsProcessInfo = await this.DWebView.EvaluateAsyncJavascriptCode($$"""
           new Promise((resolve,reject)=>{
                addEventListener("message", async event => {
                    if (event.data === "js-process/create-process") {
                        const fetch_port = event.ports[0];
                        try{
                            resolve(await createProcess(`{{env_script_url}}`,JSON.stringify({{metadata_json}}),JSON.stringify({{env_json}}),fetch_port,`{{host}}`))
                        }catch(err){
                            reject(err)
                        }
                    }
                }, { once: true })
            })
        """.Trim(), () => this.DWebView.PostMessage("js-process/create-process", new[] { port1 }));

        Console.WriteLine(String.Format("processInfo {0}", nsProcessInfo));

        var processId = (int)(NSNumber)nsProcessInfo.ValueForKey(new NSString("process_id"));
        var processInfo = new ProcessInfo(processId);
        return new ProcessHandler(processInfo, new MessagePortIpc(port2, remoteModule, IPC_ROLE.CLIENT));
    }

    public record RunProcessMainOptions(string MainUrl);

    public Task RunProcessMain(int process_id, RunProcessMainOptions options) =>
        this.DWebView.EvaluateJavaScriptAsync("void runProcessMain(" + process_id + ", {main_url:`" + options.MainUrl + "`})");

    public Task DestroyProcess(int process_id) =>
        this.DWebView.EvaluateJavaScriptAsync(String.Format("void destroy({0})", process_id).Trim());

    public async Task<int> CreateIpc(int process_id, Mmid mmid)
    {
        var channel = await this.DWebView.CreateWebMessageChannel();
        var port1 = channel.Port1;
        var port2 = channel.Port2;
        await this.DWebView.EvaluateAsyncJavascriptCode($$"""
        new Promise((resolve,reject)=>{
            addEventListener("message", async event => {
                if (event.data === "js-process/create-ipc") {
                    const ipc_port = event.ports[0];
                    try{
                        resolve(await createIpc($process_id, `$mmid`, ipc_port))
                    }catch(err){
                        reject(err)
                    }
                }
            }, { once: true })
        })
        """.Trim(), () => this.DWebView.PostMessage("js-process/create-pic", new[] { port1 }));

        return IpcWebMessageCache.SaveNative2JsIpcPort(port2);
    }

    public void Destroy()
    { }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/sys/js/JsProcessWebApi.kt
package info.bagen.dwebbrowser.microService.sys.js

import android.net.Uri
import android.webkit.WebMessage
import info.bagen.dwebbrowser.microService.helper.Mmid
import info.bagen.dwebbrowser.microService.helper.gson
import info.bagen.dwebbrowser.microService.ipc.IPC_ROLE
import info.bagen.dwebbrowser.microService.ipc.Ipc
import info.bagen.dwebbrowser.microService.ipc.ipcWeb.MessagePortIpc
import info.bagen.dwebbrowser.microService.ipc.ipcWeb.saveNative2JsIpcPort
import info.bagen.dwebbrowser.microService.webview.DWebView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicInteger

class JsProcessWebApi(val dWebView: DWebView) {
    suspend fun isReady() =
        dWebView.evaluateSyncJavascriptCode("typeof createProcess") == "function"

    data class ProcessInfo(val process_id: Int) {}
    inner class ProcessHandler(val info: ProcessInfo, var ipc: MessagePortIpc)

    /**
     * 执行js"多步骤"代码时的并发编号
     */
    private var hidAcc = AtomicInteger(1);

    suspend fun createProcess(
      env_script_url: String,
      metadata_json: String,
      env_json: String,
      remoteModule: Ipc.MicroModuleInfo,
      host: String
    ) = withContext(Dispatchers.Main) {
        val channel = dWebView.createWebMessageChannel()
        val port1 = channel[0]
        val port2 = channel[1]
        val metadata_json_str = gson.toJson(metadata_json)
        val env_json_str = gson.toJson(env_json)

        val hid = hidAcc.addAndGet(1);
        val processInfo_json = dWebView.evaluateAsyncJavascriptCode("""
            new Promise((resolve,reject)=>{
                addEventListener("message", async function doCreateProcess(event) {
                    if (event.data === "js-process/create-process/$hid") {
                        removeEventListener("message", doCreateProcess);
                        const fetch_port = event.ports[0];
                        try{
                            resolve(await createProcess(`$env_script_url`,$metadata_json_str,$env_json_str,fetch_port,`$host`))
                        }catch(err){
                            reject(err)
                        }
                    }
                })
            })
            """.trimIndent(), afterEval = {
            try {
                dWebView.postWebMessage(
                    WebMessage("js-process/create-process/$hid", arrayOf(port1)), Uri.EMPTY
                );
            } catch (e: Exception) {
                println("QQQQ");
                e.printStackTrace()
            }
        })
        debugJsProcess("processInfo", processInfo_json)
        val info = gson.fromJson(processInfo_json, ProcessInfo::class.java)
        ProcessHandler(info, MessagePortIpc(port2, remoteModule, IPC_ROLE.CLIENT))
    }


    data class RunProcessMainOptions(val main_url: String)

    suspend fun runProcessMain(process_id: Int, options: RunProcessMainOptions) =
        dWebView.evaluateAsyncJavascriptCode(
            """
        runProcessMain($process_id, { main_url:`${options.main_url}` })
        """.trimIndent()
        ).let {}

    suspend fun destroyProcess(process_id: Int) =
        dWebView.evaluateAsyncJavascriptCode(
            """
        destroyProcess($process_id)
        """.trimIndent()
        ).let {}

    suspend fun createIpc(process_id: Int, mmid: Mmid) = withContext(Dispatchers.Main) {
        val channel = dWebView.createWebMessageChannel()
        val port1 = channel[0]
        val port2 = channel[1]
        val hid = hidAcc.getAndAdd(1);
        dWebView.evaluateAsyncJavascriptCode("""
        new Promise((resolve,reject)=>{
            addEventListener("message", async function doCreateIpc(event) {
                if (event.data === "js-process/create-ipc/$hid") {
                    removeEventListener("message", doCreateIpc);
                    const ipc_port = event.ports[0];
                    try{
                        resolve(await createIpc($process_id, `$mmid`, ipc_port))
                    }catch(err){
                        reject(err)
                    }
                }
            })
        })
        """.trimIndent(), afterEval = {
            dWebView.postWebMessage(
                WebMessage("js-process/create-ipc/$hid", arrayOf(port1)), Uri.EMPTY
            );
        })
        saveNative2JsIpcPort(port2)
    }



    fun destroy() {
        dWebView.destroy()
    }
}